Metals 1.6.5 breaks with scala-cli 1.12.3

**Version(s)**
1.12.3

**Describe the bug**
Metals integration with a scala-cli project was working perfectly in nvim (via nvim-metals) with scala-cli 1.12.2 using Adoptium JDK 25.0.2 on macOS Tahoe 26.3 on a MacBook Pro M1 Max. Upgrading to scala-cli 1.12.3 (and cleaning all the scala-cli/metals/bsp artifacts) results in the following in the `.metals/metals.log`

```
❯ cat .metals/metals.log
2026.02.26 21:40:54 WARN  Using H2 2.4.240 which is newer than the version Flyway has been verified with. The latest verified version of H2 is 2.3.232.
2026.02.26 21:40:54 INFO  no build target found for /Users/djn/Git/Muve-4/cte/tools/com/araxis/versions/CheckVersions.scala. Using presentation compiler with project's scala-library version: 3.3.7
2026.02.26 21:40:55 INFO  Wrote configuration file for ide in: /Users/djn/Git/Muve-4/cte/tools/.bsp/scala-cli.json
2026.02.26 21:40:55 INFO  time: ran 'scala-cli bspConfig' in 60ms
2026.02.26 21:40:55 INFO  Attempting to connect to the build server...
2026.02.26 21:40:55 INFO  Running BSP server List(/Users/djn/Library/Application Support/Coursier/bin/.scala-cli.aux, bsp, --json-options, /Users/djn/Git/Muve-4/cte/tools/.scala-build/ide-options-v2.json, --json-launcher-options, /Users/djn/Git/Muve-4/cte/tools/.scala-build/ide-launcher-options.json, --envs-file, /Users/djn/Git/Muve-4/cte/tools/.scala-build/ide-envs.json, /Users/djn/Git/Muve-4/cte/tools)
2026.02.26 21:40:55 INFO  tracing is disabled for protocol BSP, to enable tracing of incoming and outgoing JSON messages create an empty file at /Users/djn/Git/Muve-4/cte/.metals/bsp.trace.json or /Users/djn/Library/Caches/org.scalameta.metals/bsp.trace.json
2026.02.26 21:40:55 INFO  time: Connected to build server in 0.83s
2026.02.26 21:40:56 INFO  Connected to Build server: scala-cli v1.12.3
2026.02.26 21:40:56 INFO  BSP server: Feb 26, 2026 9:40:56 PM org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint request
2026.02.26 21:40:56 INFO  BSP server: WARNING: Unsupported request method: buildTarget/wrappedSources
2026.02.26 21:40:55 INFO  Failed to run request with params WrappedSourcesResult
org.eclipse.lsp4j.jsonrpc.ResponseErrorException: Unsupported request method: buildTarget/wrappedSources
        at org.eclipse.lsp4j.jsonrpc.RemoteEndpoint.handleResponse(RemoteEndpoint.java:220)
        at org.eclipse.lsp4j.jsonrpc.RemoteEndpoint.consume(RemoteEndpoint.java:204)
        at scala.meta.internal.metals.RequestMonitorImpl$$anon$1.consume(ServerLivenessMonitor.scala:41)
        at org.eclipse.lsp4j.jsonrpc.json.StreamMessageProducer.handleMessage(StreamMessageProducer.java:185)
        at org.eclipse.lsp4j.jsonrpc.json.StreamMessageProducer.listen(StreamMessageProducer.java:97)
        at org.eclipse.lsp4j.jsonrpc.json.ConcurrentMessageProcessor.run(ConcurrentMessageProcessor.java:114)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:545)
        at java.util.concurrent.FutureTask.run(FutureTask.java:328)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1090)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:614)
        at java.lang.Thread.run(Thread.java:1474)

2026.02.26 21:40:56 INFO  Disconnecting from scala-cli session...
2026.02.26 21:40:56 ERROR Failed to connect with build server, no functionality will work.
scala.meta.internal.metals.MetalsBspException: BSP connection failed in the attempt to get: WrappedSourcesResult
        at scala.meta.internal.metals.BuildServerConnection$$anonfun$1.$anonfun$applyOrElse$9(BuildServerConnection.scala:585)
        at scala.Option.getOrElse(Option.scala:201)
        at scala.meta.internal.metals.BuildServerConnection$$anonfun$1.applyOrElse(BuildServerConnection.scala:585)
        at scala.meta.internal.metals.BuildServerConnection$$anonfun$1.applyOrElse(BuildServerConnection.scala:560)
        at scala.concurrent.impl.Promise$Transformation.run(Promise.scala:523)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1090)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:614)
        at java.lang.Thread.run(Thread.java:1474)
Caused by: org.eclipse.lsp4j.jsonrpc.ResponseErrorException: Unsupported request method: buildTarget/wrappedSources
        at org.eclipse.lsp4j.jsonrpc.RemoteEndpoint.handleResponse(RemoteEndpoint.java:220)
        at org.eclipse.lsp4j.jsonrpc.RemoteEndpoint.consume(RemoteEndpoint.java:204)
        at scala.meta.internal.metals.RequestMonitorImpl$$anon$1.consume(ServerLivenessMonitor.scala:41)
        at org.eclipse.lsp4j.jsonrpc.json.StreamMessageProducer.handleMessage(StreamMessageProducer.java:185)
        at org.eclipse.lsp4j.jsonrpc.json.StreamMessageProducer.listen(StreamMessageProducer.java:97)
        at org.eclipse.lsp4j.jsonrpc.json.ConcurrentMessageProcessor.run(ConcurrentMessageProcessor.java:114)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:545)
        at java.util.concurrent.FutureTask.run(FutureTask.java:328)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1090)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:614)
        at java.lang.Thread.run(Thread.java:1474)

2026.02.26 21:40:56 INFO  BSP server: Feb 26, 2026 9:40:56 PM org.eclipse.lsp4j.jsonrpc.RemoteEndpoint handleNotification
2026.02.26 21:40:56 INFO  BSP server: WARNING: Notification threw an exception: {
2026.02.26 21:40:56 INFO  BSP server:   "jsonrpc": "2.0",
2026.02.26 21:40:56 INFO  BSP server:   "method": "build/exit",
2026.02.26 21:40:55 INFO  Shut down connection with build server.
2026.02.26 21:40:56 INFO  BSP server:   "params": null
2026.02.26 21:40:56 INFO  BSP server: }
2026.02.26 21:40:56 INFO  BSP server: java.lang.RuntimeException: java.lang.reflect.InvocationTargetException
2026.02.26 21:40:56 INFO  BSP server:   at org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint.lambda$null$0(GenericEndpoint.java:67)
2026.02.26 21:40:56 INFO  BSP server:   at org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint.notify(GenericEndpoint.java:152)
2026.02.26 21:40:56 INFO  BSP server:   at org.eclipse.lsp4j.jsonrpc.RemoteEndpoint.handleNotification(RemoteEndpoint.java:220)
2026.02.26 21:40:56 INFO  BSP server:   at org.eclipse.lsp4j.jsonrpc.RemoteEndpoint.consume(RemoteEndpoint.java:187)
2026.02.26 21:40:56 INFO  BSP server:   at org.eclipse.lsp4j.jsonrpc.json.StreamMessageProducer.handleMessage(StreamMessageProducer.java:194)
2026.02.26 21:40:56 INFO  BSP server:   at org.eclipse.lsp4j.jsonrpc.json.StreamMessageProducer.listen(StreamMessageProducer.java:94)
2026.02.26 21:40:56 INFO  BSP server:   at org.eclipse.lsp4j.jsonrpc.json.ConcurrentMessageProcessor.run(ConcurrentMessageProcessor.java:113)
2026.02.26 21:40:56 INFO  BSP server:   at java.base@17.0.9/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539)
2026.02.26 21:40:56 INFO  BSP server:   at java.base@17.0.9/java.util.concurrent.FutureTask.run(FutureTask.java:264)
2026.02.26 21:40:56 INFO  BSP server:   at java.base@17.0.9/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
2026.02.26 21:40:56 INFO  BSP server:   at java.base@17.0.9/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
2026.02.26 21:40:56 INFO  BSP server:   at java.base@17.0.9/java.lang.Thread.run(Thread.java:840)
2026.02.26 21:40:56 INFO  BSP server:   at com.oracle.svm.core.thread.PlatformThreads.threadStartRoutine(PlatformThreads.java:807)
```

Downgrading scala-cli to 1.12.2 fixes the problem.

**To Reproduce**
1. Upgrade to scala-cli 1.12.3.
2. In nvim (or, likely, other LSP editors/IDEs), try to make Metals work with a scala-cli project.

**Expected behaviour**
Metals 1.6.5 should continue to work with scala-cli projects after upgrading to scala-cli 1.12.3.

**Suggested root cause**
Claude Code has summarized what it believes to be the cause of the problem as follows:

> Metals 1.6.5 sends a buildTarget/wrappedSources BSP request to the scala-cli BSP server during connection initialisation. This method is defined in the ScalaScriptBuildServer Java interface (modules/scala-cli-bsp/src/main/java/scala/build/bsp/ScalaScriptBuildServer.java) and annotated with @JsonRequest("buildTarget/wrappedSources"). The server-side implementation exists in BspServer.scala and is proxied through BuildServerProxy.scala.

> In the v1.12.3 native image, LSP4J's GenericEndpoint fails to discover this method and responds with 'Unsupported request method: buildTarget/wrappedSources'. This suggests that the GraalVM native image reflection configuration does not include the ScalaScriptBuildServer interface, so LSP4J cannot reflectively find the @JsonRequest-annotated method at runtime.

> Metals treats the failure of this request as fatal (there is no fallback handler for wrappedSources in BuildServerConnection.scala), so the entire BSP connection is torn down and no IDE functionality works.

--- a/modules/cli/src/main/resources/META-INF/native-image/org.virtuslab/scala-cli-core/reflect-config.json
+++ b/modules/cli/src/main/resources/META-INF/native-image/org.virtuslab/scala-cli-core/reflect-config.json
@@ -1243,6 +1243,42 @@
     "allDeclaredFields": true
   },
   {
+    "name": "scala.build.bsp.ScalaScriptBuildServer",
+    "queryAllDeclaredMethods": true,
+    "allDeclaredConstructors": true,
+    "allPublicConstructors": true,
+    "allDeclaredMethods": true,
+    "allDeclaredFields": true
+  },
+  {
+    "name": "scala.build.bsp.WrappedSourceItem",
+    "allDeclaredConstructors": true,
+    "allPublicConstructors": true,
+    "allDeclaredMethods": true,
+    "allDeclaredFields": true
+  },
+  {
+    "name": "scala.build.bsp.WrappedSourcesItem",
+    "allDeclaredConstructors": true,
+    "allPublicConstructors": true,
+    "allDeclaredMethods": true,
+    "allDeclaredFields": true
+  },
+  {
+    "name": "scala.build.bsp.WrappedSourcesParams",
+    "allDeclaredConstructors": true,
+    "allPublicConstructors": true,
+    "allDeclaredMethods": true,
+    "allDeclaredFields": true
+  },
+  {
+    "name": "scala.build.bsp.WrappedSourcesResult",
+    "allDeclaredConstructors": true,
+    "allPublicConstructors": true,
+    "allDeclaredMethods": true,
+    "allDeclaredFields": true
+  },
+  {
     "name": "scala.build.bsp.protocol.TextEdit",
     "allDeclaredConstructors": true,
     "allPublicConstructors": true,

--- a/modules/integration/src/test/java/scala/cli/integration/bsp/WrappedSourceItem.java
+++ b/modules/integration/src/test/java/scala/cli/integration/bsp/WrappedSourceItem.java
@@ -1,0 +1,8 @@
+package scala.cli.integration.bsp;
+
+public class WrappedSourceItem {
+  public String uri;
+  public String generatedUri;
+  public String topWrapper;
+  public String bottomWrapper;
+}

--- a/modules/integration/src/test/java/scala/cli/integration/bsp/WrappedSourcesItem.java
+++ b/modules/integration/src/test/java/scala/cli/integration/bsp/WrappedSourcesItem.java
@@ -1,0 +1,9 @@
+package scala.cli.integration.bsp;
+
+import ch.epfl.scala.bsp4j.BuildTargetIdentifier;
+import java.util.List;
+
+public class WrappedSourcesItem {
+  public BuildTargetIdentifier target;
+  public List<WrappedSourceItem> sources;
+}

--- a/modules/integration/src/test/java/scala/cli/integration/bsp/WrappedSourcesResult.java
+++ b/modules/integration/src/test/java/scala/cli/integration/bsp/WrappedSourcesResult.java
@@ -1,0 +1,7 @@
+package scala.cli.integration.bsp;
+
+import java.util.List;
+
+public class WrappedSourcesResult {
+  public List<WrappedSourcesItem> items;
+}

--- a/modules/integration/src/test/scala/scala/cli/integration/BspSuite.scala
+++ b/modules/integration/src/test/scala/scala/cli/integration/BspSuite.scala
@@ -79,7 +79,8 @@ trait BspSuite { this: ScalaCliSuite =>
     f: (
       os.Path,
       TestBspClient,
-      b.BuildServer & b.ScalaBuildServer & b.JavaBuildServer & b.JvmBuildServer
+      b.BuildServer & b.ScalaBuildServer & b.JavaBuildServer & b.JvmBuildServer &
+        TestBspClient.WrappedSourcesBuildServer
     ) => Future[T]
   ): T = withBspInitResults(
     inputs,
@@ -107,7 +108,8 @@ trait BspSuite { this: ScalaCliSuite =>
     f: (
       os.Path,
       TestBspClient,
-      b.BuildServer & b.ScalaBuildServer & b.JavaBuildServer & b.JvmBuildServer,
+      b.BuildServer & b.ScalaBuildServer & b.JavaBuildServer & b.JvmBuildServer &
+        TestBspClient.WrappedSourcesBuildServer,
       b.InitializeBuildResult
     ) => Future[T]
   ): T = {
@@ -120,8 +122,9 @@ trait BspSuite { this: ScalaCliSuite =>
 
       val proc = os.proc(TestUtil.cli, "bsp", bspOptions ++ extraOptionsOverride, args)
         .spawn(cwd = root, stderr = stderr, env = bspEnvs)
-      var remoteServer: b.BuildServer & b.ScalaBuildServer & b.JavaBuildServer & b.JvmBuildServer =
-        null
+      var remoteServer
+        : b.BuildServer & b.ScalaBuildServer & b.JavaBuildServer & b.JvmBuildServer &
+          TestBspClient.WrappedSourcesBuildServer = null
 
       val bspServerExited = Promise[Unit]()
       val t               = new Thread("bsp-server-watcher") {

--- a/modules/integration/src/test/scala/scala/cli/integration/BspTestDefinitions.scala
+++ b/modules/integration/src/test/scala/scala/cli/integration/BspTestDefinitions.scala
@@ -2394,4 +2394,55 @@ abstract class BspTestDefinitions extends ScalaCliSuite
       }
     }
   }
+
+  test("buildTarget/wrappedSources") {
+    val inputs = TestInputs(
+      os.rel / "simple.sc" ->
+        s"""val msg = "Hello"
+           |println(msg)
+           |""".stripMargin
+    )
+
+    withBsp(inputs, Seq(".")) { (root, _, remoteServer) =>
+      Future {
+        val buildTargetsResp = remoteServer.workspaceBuildTargets().asScala.await
+        val target           = {
+          val targets = buildTargetsResp.getTargets.asScala.map(_.getId).toSeq
+          expect(targets.length == 2)
+          extractMainTargets(targets)
+        }
+
+        val targets = List(target).asJava
+
+        val compileResp =
+          remoteServer.buildTargetCompile(new b.CompileParams(targets)).asScala.await
+        expect(compileResp.getStatusCode == b.StatusCode.OK)
+
+        val resp = remoteServer
+          .buildTargetWrappedSources(new b.SourcesParams(targets))
+          .asScala
+          .await
+
+        val items = resp.items.asScala
+        expect(items.nonEmpty)
+
+        val mainItem = items.find(_.target.getUri == target.getUri).get
+
+        val sources = mainItem.sources.asScala
+        expect(sources.size == 1)
+
+        val wrappedSource = sources.head
+        val sourceUri     = TestUtil.normalizeUri(wrappedSource.uri)
+        val expectedUri   =
+          TestUtil.normalizeUri((root / "simple.sc").toNIO.toUri.toASCIIString)
+        expect(sourceUri == expectedUri)
+
+        expect(wrappedSource.topWrapper.contains("simple"))
+        expect(wrappedSource.topWrapper.contains("scriptPath"))
+        expect(wrappedSource.bottomWrapper == "}")
+        expect(wrappedSource.generatedUri.endsWith(".scala"))
+        expect(wrappedSource.generatedUri.contains("simple"))
+      }
+    }
+  }
 }

--- a/modules/integration/src/test/scala/scala/cli/integration/TestBspClient.scala
+++ b/modules/integration/src/test/scala/scala/cli/integration/TestBspClient.scala
@@ -4,8 +4,9 @@ import ch.epfl.scala.bsp4j as b
 import org.eclipse.lsp4j.jsonrpc
 
 import java.io.{InputStream, OutputStream}
-import java.util.concurrent.ExecutorService
+import java.util.concurrent.{CompletableFuture, ExecutorService}
 
+import scala.cli.integration.bsp.WrappedSourcesResult
 import scala.collection.mutable
 import scala.concurrent.{ExecutionContext, Future, Promise}
 import scala.util.{Failure, Success}
@@ -100,8 +101,14 @@ class TestBspClient extends b.BuildClient {
 
 object TestBspClient {
 
+  trait WrappedSourcesBuildServer {
+    @org.eclipse.lsp4j.jsonrpc.services.JsonRequest("buildTarget/wrappedSources")
+    def buildTargetWrappedSources(params: b.SourcesParams)
+      : CompletableFuture[WrappedSourcesResult]
+  }
+
   private trait BuildServer extends b.BuildServer with b.ScalaBuildServer with b.JavaBuildServer
-      with b.JvmBuildServer
+      with b.JvmBuildServer with WrappedSourcesBuildServer
 
   def connect(
     in: InputStream,
@@ -109,7 +116,8 @@ object TestBspClient {
     es: ExecutorService
   ): (
     TestBspClient,
-    b.BuildServer & b.ScalaBuildServer & b.JavaBuildServer & b.JvmBuildServer,
+    b.BuildServer & b.ScalaBuildServer & b.JavaBuildServer & b.JvmBuildServer &
+      WrappedSourcesBuildServer,
     Future[Unit]
   ) = {
 

